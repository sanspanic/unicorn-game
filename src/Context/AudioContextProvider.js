import React, { useRef } from "react";
import AudioContext from "./AudioContext";
import useSound from "use-sound";
import Bump from "../Assets/Audio/bump.wav";
import Move from "../Assets/Audio/move.wav";
import Lose from "../Assets/Audio/lose.wav";
import Win from "../Assets/Audio/win.wav";

const AudioContextProvider = ({ children }) => {
  //plays on illegal move
  const [playIllegal] = useSound(Bump, { volume: 0.25 });

  //plays on legal move
  const [playLegal] = useSound(Move, { volume: 0.25 });

  //plays on game over
  const [playGameOver] = useSound(Lose, { volume: 0.25 });

  //plays on game won
  const [playGameWon] = useSound(Win, { volume: 0.25 });

  return (
    <AudioContext.Provider
      value={{ playIllegal, playLegal, playGameOver, playGameWon }}
    >
      {children}
    </AudioContext.Provider>
  );
};

export default AudioContextProvider;
