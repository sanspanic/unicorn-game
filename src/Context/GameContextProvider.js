import React, { useState } from "react";
import GameContext from "./GameContext";

const GameContextProvider = ({ children }) => {
  const initialState = {
    status: "won",
    difficulty: 0,
    size: 15,
    name: "",
  };
  const [gameData, setGameData] = useState(initialState);

  //restarts game by setting gameData.status back to "setup" & redirecting to homepage
  const playAgain = () => {
    setGameData({ ...gameData, status: "setup" });
  };

  return (
    <GameContext.Provider value={{ gameData, setGameData, playAgain }}>
      {children}
    </GameContext.Provider>
  );
};

export default GameContextProvider;
