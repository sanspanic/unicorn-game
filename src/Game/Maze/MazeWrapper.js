import React, { useContext } from "react";
import GameContext from "../../Context/GameContext";
import MazeGrid from "./MazeGrid";

const MazeWrapper = () => {
  const { gameData } = useContext(GameContext);
  return (
    <div>
      <h2 className="text-center py-10 font-mono">
        Help {gameData.name} reach the rainbow!{" "}
      </h2>
      <div className="mx-auto sm:w-9/12 md:w-6/12 lg:w-5/12 xl:w-4/12">
        <MazeGrid />
      </div>
    </div>
  );
};

export default MazeWrapper;
