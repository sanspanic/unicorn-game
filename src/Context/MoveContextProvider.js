import React, { useState, useContext, useEffect } from "react";
import MoveContext from "./MoveContext";
import UnicornApi from "../API/UnicornApi";
import GameContext from "./GameContext";
import { getSpritePositions } from "../Game/Maze/mazeHelpers";

const MoveContextProvider = ({ children }) => {
  const [spritePositions, setSpritePositions] = useState({});
  const { gameData, setGameData } = useContext(GameContext);
  const [mazeId, setMazeId] = useState("");

  //sets focus on maze grid to enable hotkeys
  //unfortunately useRef didn't work so I'm using, inappropriately, useEffect
  useEffect(() => {
    if (mazeId === "") return;
    console.log("SELECTING DOM ELEMENT TO FOCUS, mazeId: ", mazeId);
    const Maze = document.querySelector("[tabindex='-1']");
    Maze.focus();
  }, [mazeId]);

  //handles move
  const handleMove = (direction) => {
    const move = async (data) => {
      try {
        const res = await UnicornApi.makeMove(data, mazeId);
        console.log("MADE A MOVE: ", res);
        //if (res.state === "won") setGameState("won");
        if (res["state-result"] === "Move accepted") {
          updatePositions();
        } else {
          //TODO: handle illegal move
        }
        checkForGameOver(res);
      } catch (e) {
        console.log(e);
      }
    };
    move({ direction });
  };

  //updates positions of sprites following a move
  const updatePositions = async () => {
    try {
      const res = await UnicornApi.getMazeCurrentState(mazeId);
      setSpritePositions(getSpritePositions(res));
      console.log("UPDATED SPRITE POSITIONS AFTER MOVE: ", spritePositions);
    } catch (e) {
      console.log(e);
    }
  };

  //checks if game is won or lost
  const checkForGameOver = (res) => {
    if (res.state === "won") {
      setGameData({ ...gameData, status: "won" });
    } else if (res.state === "over") {
      setGameData({ ...gameData, status: "over" });
    }
  };

  return (
    <MoveContext.Provider
      value={{
        handleMove,
        updatePositions,
        checkForGameOver,
        mazeId,
        setMazeId,
        spritePositions,
        setSpritePositions,
      }}
    >
      {children}
    </MoveContext.Provider>
  );
};

export default MoveContextProvider;
