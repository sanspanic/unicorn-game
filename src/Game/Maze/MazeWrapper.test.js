import React from "react";
import { render } from "@testing-library/react";
import MazeWrapper from "./MazeWrapper";
import GameContextProvider from "../../Context/GameContextProvider";
import AudioContextProvider from "../../Context/AudioContextProvider";
import MoveContextProvider from "../../Context/MoveContextProvider";

test("renders without crashing", () => {
  render(
    <GameContextProvider>
      <AudioContextProvider>
        <MoveContextProvider>
          <MazeWrapper />
        </MoveContextProvider>
      </AudioContextProvider>
    </GameContextProvider>
  );
});

test("matches snapshot", () => {
  const { asFragment } = render(
    <GameContextProvider>
      <AudioContextProvider>
        <MoveContextProvider>
          <MazeWrapper />
        </MoveContextProvider>
      </AudioContextProvider>
    </GameContextProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});
