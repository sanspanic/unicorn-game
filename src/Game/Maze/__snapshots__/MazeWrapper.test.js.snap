// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-clouds md:h-screen font-mono text-center"
  >
    <h2
      class="pt-10 "
    >
      Help 
      <span
        class="text-pink-600"
      />
       reach the rainbow! Move using arrow keys, WASD or buttons.
    </h2>
    <p
      class="text-sm  pb-10"
    >
      <svg
        class="inline "
        fill="currentColor"
        height="24"
        viewBox="0 0 256 256"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <polyline
          fill="none"
          points="80 152 32 104 80 56"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
        <path
          d="M224,200a96,96,0,0,0-96-96H32"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
       Back to 
      <button
        class="underline text-pink-600"
      >
        setup
      </button>
       
    </p>
    <div
      class="mx-auto p-3 sm:w-9/12 md:w-7/12 lg:w-5/12 xl:w-4/12"
    >
      <div
        class="outline-none"
        tabindex="-1"
      >
        <div
          class="flex justify-center mb-2"
        >
          <button
            class="mx-4 gradient-pink border-2 shadow-xl border-purple-600 text-white px-3 py-2 rounded bg-blue-400 p-1"
          >
            <svg
              fill="currentColor"
              height="24"
              viewBox="0 0 256 256"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="none"
                height="256"
                width="256"
              />
              <polygon
                fill="none"
                points="136 32 232 128 136 224 136 176 104 176 104 80 136 80 136 32"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
              />
              <line
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
                x1="40"
                x2="40"
                y1="176"
                y2="80"
              />
              <line
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
                x1="72"
                x2="72"
                y1="176"
                y2="80"
              />
            </svg>
          </button>
          <button
            class="mx-4 gradient-pink border-2 shadow-xl border-purple-600 text-white px-3 py-2 rounded bg-blue-400 p-1"
          >
            <svg
              fill="currentColor"
              height="24"
              viewBox="0 0 256 256"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="none"
                height="256"
                width="256"
              />
              <polygon
                fill="none"
                points="120 32 24 128 120 224 120 176 152 176 152 80 120 80 120 32"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
              />
              <line
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
                x1="216"
                x2="216"
                y1="176"
                y2="80"
              />
              <line
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
                x1="184"
                x2="184"
                y1="176"
                y2="80"
              />
            </svg>
          </button>
          <button
            class="mx-4 gradient-pink border-2 shadow-xl border-purple-600 text-white px-3 py-2 rounded bg-blue-400 p-1"
          >
            <svg
              fill="currentColor"
              height="24"
              viewBox="0 0 256 256"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="none"
                height="256"
                width="256"
              />
              <polygon
                fill="none"
                points="32 136 128 232 224 136 176 136 176 104 80 104 80 136 32 136"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
              />
              <line
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
                x1="176"
                x2="80"
                y1="40"
                y2="40"
              />
              <line
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
                x1="176"
                x2="80"
                y1="72"
                y2="72"
              />
            </svg>
          </button>
          <button
            class="mx-4 gradient-pink border-2 shadow-xl border-purple-600 text-white px-3 py-2 rounded bg-blue-400 p-1"
          >
            <svg
              fill="currentColor"
              height="24"
              viewBox="0 0 256 256"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="none"
                height="256"
                width="256"
              />
              <polygon
                fill="none"
                points="32 120 128 24 224 120 176 120 176 152 80 152 80 120 32 120"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
              />
              <line
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
                x1="176"
                x2="80"
                y1="216"
                y2="216"
              />
              <line
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="16"
                x1="176"
                x2="80"
                y1="184"
                y2="184"
              />
            </svg>
          </button>
        </div>
        <div
          class="Maze-grid mx-auto text-center bg-white bg-opacity-80"
          style="grid-template-columns: repeat(15, minmax(0, 1fr));"
        />
      </div>
      <div
        class="flex justify-between"
      >
        <span>
          Difficulty: 0
        </span>
        <span>
          Size: 15
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
